sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -613,13 +633,62 @@ def _blocking_request(self, request):
     if response.error:
       raise RuntimeError(response.error)
     else:
+      # TODO mxm here we could update the cache token
       return response
 
   def _next_id(self):
     self._last_id += 1
     return str(self._last_id)
 
 
+class CachingGrpcStateHandler(GrpcStateHandler):
+  """ TODO mxm """
+
+  def __init__(self, state_stub, global_state_cache):
+    super(CachingGrpcStateHandler, self).__init__(state_stub)
+    self._state_cache = global_state_cache
+
+  def blocking_get(self, state_key, coder):
+    cache_tokens = self._context.cache_tokens
+    if not cache_tokens:
+      # no cache tokens, can't do a lookup in the cache
+      return super(CachingGrpcStateHandler, self).blocking_get(
+          state_key, coder)
+    cache_state_key = CachingGrpcStateHandler.convert_to_cache_key(state_key)
+    value = self._state_cache.get(cache_state_key, cache_tokens)
+    if not value:
+      # Cache miss, need to retrieve from the Runner
+      value = super(CachingGrpcStateHandler, self).blocking_get(
```

Comment:
We can another TODO here for adding metrics in a separate PR.

Reply:
Good point.